Develop

EvgeniyaKuzmina opened the pull request.

EvgeniyaKuzmina committed:
feat: создано два модуля: main-service и stat-service.
1. main-service:
- создан пакет user — отвечает за работу с пользователями: создан объект пользователя User.java, созданы DTO объекты пользователей, слой репозитория для работы с БД, сервисная логика и логика взаимодействия с пользователем (контролеры)

- создан пакет request — отвечает за работу с запросами на участие в событиях: создан объект запросов на участие Request.java, созданы объекты DTO запросов на участие

- создан пакет event — отвечает за работу с событиями: создан объект событий Event.java, созданы DTO объекты событий. Включает в себя пакет comment, который отвечает за работу с комментариями по событиям. Включает в себя location — отвечает за работу с локацией, где будет проходить событие

- создан пакет compilation — отвечает за работу с подборками событий: создан объект подборки Compilation.java, созданы DTO объекты подборки событий

- создан пакет category — отвечает за работу с категориями событий: создан объект категорий Category.java, созданы DTO объекты категорий, слой репозитория для работы с БД, сервисная логика и логика взаимодействия с категориями событий (контролеры)

- создан пакет status — хранит статусы для событий и статусы для запросов на участие в событиях

- создан пакет exception — хранит исключения

- создан пакет error — хранит классы для обработки исключений

EvgeniyaKuzmina committed:
feat: main-service:
- пакет request: создан слой репозитория для работы с БД, реализована сервисная логика и логика взаимодействия с пользователем (контролеры)

- создан пакет compilation: создан слой репозитория для работы с БД, реализована сервисная логика и логика взаимодействия с пользователем (контролеры)

EvgeniyaKuzmina committed:
feat: main-service:
- пакет request: создан дополнительные методы для получения запросов

- пакет event: создан слой репозитория для работы с БД, частично реализована сервисная логика

EvgeniyaKuzmina committed:
feat: main-service:

- пакет event: частично создан слой контроллера для публичного API

EvgeniyaKuzmina committed:
feat: main-service:

- пакет event: создан слой контроллера для публичного API, для приватного API и API администратора

EvgeniyaKuzmina committed:
feat: main-service:

- доработан основной функционал main-service. Пакет проходит все тесты, выполняются все запросы из спецификации

EvgeniyaKuzmina committed:
feat: stats-service:

- разработан функционал по работе со статистикой. Добавлены репозиторий, сервисный слой и слой контроллера

EvgeniyaKuzmina committed:
style: исправлено согласно требованиям code style

EvgeniyaKuzmina committed:
style: исправлено согласно требованиям code style

EvgeniyaKuzmina committed:
style: исправлено согласно требованиям code style

EvgeniyaKuzmina committed:
feat: доработан сервис статистики
Добавлен client  в main-service для работы со статистикой

EvgeniyaKuzmina committed:
fix: исправлена ошибка в названии файла sql с запросом на создание таблицы в main-service

EvgeniyaKuzmina committed:
style: удалены лишние импорты

EvgeniyaKuzmina committed:
style: изменён формат для времени в модуле статистики

EvgeniyaKuzmina committed:
feat: доделан функционал по работе со статистикой. При каждом запросе события, запрашиваются данные по просмотрам из сервиса статистики и добавляются в соответствующее поле события
style: описан функционал приложения в файле Readme

EvgeniyaKuzmina committed:
Update README.md

Добавлена ссылка на pull requests

EvgeniyaKuzmina committed:
style: удалены лишние обработчики try - catch

Frilock commented on ewm-service/pom.xml:
```diff
@@ -0,0 +1,119 @@
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>42.5.0</version>
```
Общий комментарий по pom файлам. Смотрите, у Вас многомодульный проект, сейчас он состоит из 2х модулей, но возможно их будет больше. Так вот, хочется чтобы в приложении была одна точка добавления зависимостей. Такой обычно делают корневой pom файл. Там нужно сделать блок dependencyManager и добавлять те библиотеки которых нет в parent, а затем уже в других pom файлах подключать, но уже без указания версии. Кроме того, версии этих библиотек правильнее указывать в properties, поскольку pom файлы бывают довольно больших размеров и изменение версии может быть не такой простой задачей.

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/CategoryRepository.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * класс репозиторий для работы с БД категорий событий
+ */
+
```
Лишняя пустая строчка

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/CategoryRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    // получение списка категорий с пагинацией
```
Такие комментарии не имеют смысла, по идее по названию метода должно быть понятно, что он делает, если же все таки хотите оставлять такие комментарии, то используйте javadoc - 
/**
*
*/

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/CategoryService.java:
```diff
@@ -0,0 +1,31 @@
+ */
+
+
+@Service
```
Аннотации, которые говорят о том, что класс является бином, нужно ставить над реализацией интерфейса. Над интерфейсами нет смысла в проставлении таких аннотаций

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+
+        repository.deleteById(id);
+        log.warn("CategoryServiceImpl: removeCategory — Категория с указанным eventId {} удалена", id);
+
```
Лишняя пустая строчка в конце метода, проверьте другие места пожалуйста

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/controller/CategoryAdminController.java:
```diff
@@ -0,0 +1,57 @@
+    @PostMapping
+    public CategoryDto createCategory(@Valid @RequestBody NewCategoryDto categoryDto) {
+        Category category = CategoryMapper.toCategoryFromNewCategoryDto(categoryDto);
+        log.warn(category.toString());
```
Предлагаю сделать какой то осмысленный лог, просто вот Вы выведете категорию, и что дальше? Хотя бы нужно указать в логе, какой метод был вызван. Кроме того, по идее toString() можно не указывать, он вызовется и так

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/controller/CategoryPublicController.java:
```diff
@@ -0,0 +1,56 @@
+
+@RestController
+@RequestMapping(path = "/categories")
+@Slf4j
```
Лишняя аннотация, не вижу использования логгера в данном классе

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/model/Category.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
```
Данную аннотацию можно не ставить над идентификатором

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/category/model/Category.java:
```diff
@@ -0,0 +1,43 @@
+    @Column(name = "id")
+    private Long id;
+    @NotNull
+    @Column(name = "name")
```
Смотрите, общий комментарий по скрипту и модели данных. Вы когда архитектуру БД планируете, нужно продумать размерность полей, то есть, например, поле код не должно иметь стандартный размер в 255, а его можно сделать, например, в 50 символов, а вот для описание нужно больше места, поэтому логичнее под данную колонку выделить, например, 1000. 
Кроме того, после того как Вы все указали в скрипте, тоже самое нужно сделать в аннотации Column для классов моделей. Указать размерность поля, а также дополнительные флаги, например, nullable = false.

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/client/BaseClient.java:
```diff
@@ -0,0 +1,70 @@
+
+
+    private static ResponseEntity<Object> prepareGatewayResponse(ResponseEntity<Object> response) {
+        if (response.getStatusCode().is2xxSuccessful()) {
```
А getStatusCode null вернуть может?

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+    @Override
+    public Compilation createCompilation(Compilation compilation) {
+
+        //Collection<Event> events = eventService.getAllEventByIds(compilation.getEvents());
```
Лучше закомментированный код не оставлять в проекте, через какое то время становится не ясно, для чего он

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/controller/CompilationAdminController.java:
```diff
@@ -0,0 +1,83 @@
+    public void removeCompilation(@PathVariable Long id) {
+        service.removeCompilation(id);
+    }
+
```
Заметил, что вы в конце своих классов оставляете несколько пустых строк, мелочь конечно, но лично я бы посоветовал больше одной не оставлять, это в глаза бросается =)

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/controller/CompilationAdminController.java:
```diff
@@ -0,0 +1,83 @@
+
+@RestController
+@RequestMapping(path = "/admin/compilations")
+@Slf4j
```
Лишняя аннотация

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/controller/CompilationPublicController.java:
```diff
@@ -0,0 +1,62 @@
+
+@RestController
+@RequestMapping(path = "/compilations")
+@Slf4j
```
Лишняя аннотация

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/controller/CompilationPublicController.java:
```diff
@@ -0,0 +1,62 @@
+        Pageable pageable = PageRequest.of(page, size);
+        Collection<CompilationDto> allCompilationDto = new ArrayList<>();
+        Collection<Compilation> allCompilation;
+        if (pinned == null) {
```
Мне кажется данное выражение можно заменить на тернарный оператор, будет компактнее

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/compilation/mapper/CompilationMapper.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Collection;
+import java.util.List;
+
+@Slf4j
```
Лишняя аннотация

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,341 @@
+
+        log.info("EventServiceImpl: createEvent — Добавлено событие {}.", event);
+        return repository.save(event);
+
```
Лишняя строчка в конце метода

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,341 @@
+        validateUserIdAndEventId(updEvent.getId(), userId);
+        Event event = getEventById(updEvent.getId());
+        LocalDateTime publishedTime = LocalDateTime.now();
+        if (event.getState().equals(Status.CANCELED)) {
```
enum поля правильнее сравнивать через стандартные операторы != и ==, работать будет в любом случае, но стандартные операторы как минимум быстрее, кроме того не будет NPE.
Вообще не знаю, говорили Вам или нет, но советую установить в IDEA плагин sonarlint - статический анализатор кода, который покажет ряд хороших практик и возможные места с ошибками

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,341 @@
+            log.error("EventServiceImpl: updateEventByInitiator — Событие изменить нельзя");
+            throw new ConflictException("Событие изменить нельзя");
+        }
+        if (event.getEventDate().isBefore(publishedTime.plusHours(2))) {
```
А event,getEventDate null не вернет?

Frilock replied:
Тут не ответили

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventAdminController.java:
```diff
@@ -0,0 +1,107 @@
+
+
+    // преобразование параметров запроса в объект EventParam
+    private EventParam creatParam(List<Long> usersId,
```
Опечатка в слове create

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventAdminController.java:
```diff
@@ -0,0 +1,107 @@
+
+@RestController
+@RequestMapping(path = "/admin/events")
+@Slf4j
```
Лишняя аннотация, дальше писать о ней не буду, проверьте пожалуйста дальше сами

EvgeniyaKuzmina replied:
исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventPrivateController.java:
```diff
@@ -0,0 +1,133 @@
+        Category category = categoryService.getCategoryById(eventDto.getCategory());
+        Event event = EventMapper.toEventFromNewDto(eventDto, category);
+        LocalDateTime publishedTime = LocalDateTime.now();
+        if (event.getEventDate().isBefore(publishedTime.plusHours(2))) {
```
Кажется что тут может event.getEventDate() вернуться null и будет NPE, поправьте если не прав

Frilock replied:
Не ответили на мой комментарий

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/dto/EventShortDto.java:
```diff
@@ -0,0 +1,45 @@
+@Getter
+@Setter
+public class EventShortDto {
+    @Id
```
Мне казалось обычно, Id аннотацию используют только в классах моделей, зачем она тут?

EvgeniyaKuzmina replied:
Вставляла копи пастом, видимо не удалила.
Исправлено

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/model/Event.java:
```diff
@@ -0,0 +1,98 @@
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Event event = (Event) o;
+        return id.equals(event.id) && title.equals(event.title) && description.equals(event.description) && annotation.equals(event.annotation) && initiator.equals(event.initiator) && category.equals(event.category);
```
Слишком длинная строчка, давайте сделаем переносы по && символу

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/request/RequestServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+                .requester(user)
+                .event(event)
+                .build();
+        if (event.getRequestModeration().equals(false)) {
```
Лучше использовать вместо false - Boalean.FALSE

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/request/RequestServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+        // отменяем все заявки в статусе ожидания, если при подтверждении текущей заявки лимит заявок исчерпан
+        if (event.getParticipantLimit().equals(confirmedRequests)) {
+            List<Request> requests = getRequestsByEventIdAndStatus(event.getId(), Status.PENDING);
+            for (Request r : requests) {
```
Мне кажется тут не совсем оптимальная структура, вы в цикле делаете запросы на отклонение заявки. А мы разве не можем собрать коллекцию заявок на отклонение и одним запросом им обновить статусы

Frilock commented on ewm-service/src/main/resources/application.properties:
```diff
@@ -0,0 +1,30 @@
+spring.datasource.password=test
+spring.h2.console.enabled=true
+
+
```
Куча лишних пустых строчек в конце файла давайте их уберем

Frilock commented on ewm-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,103 @@
+CREATE TABLE IF NOT EXISTS comments
+(
+    id        BIGINT GENERATED BY DEFAULT AS IDENTITY NOT NULL,
+    text      VARCHAR(10000)                          NOT NULL,
```
Слишком большое поле на мой взгляд,, Postgres поддерживаем такую размерность? Я честно говоря сам запамятовал

Frilock commented on ewm-service/src/test/java/ru/yandex/practicum/mainservice/MainServerApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+class MainServerApplicationTests {
+
+    @Test
+    void contextLoads() {
```
Если честно, то хотелось бы от Вас увидеть еще и тесты на Ваши сервисы (unit тесты), знаю, что это задание со звездочкой, но вроде бы у Вас еще есть время. Кроме того, может быть сами какие то неточности отловите

Frilock commented on stats-server/src/main/java/ru/yandex/practicum/statserver/statistic/StatisticMapper.java:
```diff
@@ -0,0 +1,18 @@
+import ru.yandex.practicum.statserver.statistic.dto.EndpointHit;
+import ru.yandex.practicum.statserver.statistic.model.Statistic;
+
+@Slf4j
```
Лишняя аннотация

Frilock commented on stats-server/src/main/java/ru/yandex/practicum/statserver/statistic/StatisticRepository.java:
```diff
@@ -0,0 +1,41 @@
+public interface StatisticRepository extends JpaRepository<Statistic, Long> {
+
+    // получение информации о просмотрах с уникальным ip
+    @Query("select new ru.yandex.practicum.statserver.statistic.dto.ViewStats(s.app, s.uri, count(distinct s.ip)) " +
```
А вы не тестировали, ему разве весь путь нужен такой ru.yandex.practicum.statserver.statistic.dto.?

Frilock commented on stats-server/src/main/java/ru/yandex/practicum/statserver/statistic/StatisticServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public Collection<ViewStats> getStatistic(LocalDateTime start, LocalDateTime end, Collection<String> uris, Boolean unique) {
+        if (unique && uris != null)
```
Хотел бы Вас попросить все таки обособлять данные конструкции даже если они состоят из 1 строчки, дабы защитить себя от ошибок по невнимательности

Frilock reviewed (changes requested):
Добрый день, оставил ряд комментариев к Вашей работе. Если будут какие то вопросы приходите ко мне в Slack, по возможности постараюсь на них ответить.

EvgeniyaKuzmina committed:
style: исправлены комментарии по оформлению кода: удалены лишние строки, комментарии вида // заменены на комментарии по javadoc

EvgeniyaKuzmina committed:
style: исправлены комментарии по оформлению кода: удалены лишние строки, комментарии вида // заменены на комментарии по javadoc, удалены лишние аннотации в методах

EvgeniyaKuzmina committed:
style: исправлены зависимости в pom файлах. Все зависимости указаны в родительском файле с указанием версии, номера версии прописаны в properties. В pom файлах наследниках версии зависимостей удалены.

EvgeniyaKuzmina committed:
style: исправлены зависимости в pom файлах. Удалены версии всех указанных зависимостей. Все используемые зависимости в проекте есть в родительском springframework

EvgeniyaKuzmina committed:
style: исправлены стилистически ошибки в написании кода: удалены лишние пустые строки и комментарии. Указаны дополнительно размерность поле у объектов связанных с БД. Исправлена грамматическая ошибка в названии метода.

EvgeniyaKuzmina committed:
style: исправлены условия проверки булевых значений в методах. Теперь сравнение происходит с константами Boolean.TRUE и Boolean.FALSE.
Изменена логика изменения статуса всем заявка на отклонённые, если количество мест на участие закончилось. Изменение статуса теперь происходит через запрос в БД.
Удалены лишние пробелы

EvgeniyaKuzmina committed:
style: изменён метод отменяющий все заявки определённого события. Теперь получение списка всех заявок по параметрам и изменение статуса происходит через запрос в БД

EvgeniyaKuzmina committed:
style: удалён лишний импорт в классе

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/client/BaseClient.java:
```diff
@@ -0,0 +1,70 @@
+
+        ResponseEntity<Object> serverResponse;
+        try {
+            if (parameters != null) {
```
Тут тоже по идее можно на тернарное выражение было заменить

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/client/EventClient.java:
```diff
@@ -0,0 +1,91 @@
+ * класс для отправки запросов в проект статистики
+ */
+@Service
+@Slf4j
```
Опять же лишняя аннотация

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/client/EventClient.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private Collection<ViewStats> parseResponseEntityToViewStats(ResponseEntity<Object> response) {
+        Gson gson = new Gson();
```
А зачем создавать обьект gson на каждый вызов метода? Можно его как поле класса создать по идее

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/dto/NewEventDto.java:
```diff
@@ -0,0 +1,45 @@
+    @Length(min = 3)
+    private String title;
+    @NotNull
+    @Length(max = 7000)
```
Вроде обе Length можно в одну уместить, только с 2мя парметрами

Frilock commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/location/Location.java:
```diff
@@ -0,0 +1,27 @@
+@Setter
+@ToString
+@Entity
+@Table(name = "locations", schema = "public")
```
public дефолтная схема, ее можно не указывать

Frilock reviewed (changes requested):
Добрый день, почти все мои комментарии поправили, написал еще пару вопросов, и можно будет принимать

EvgeniyaKuzmina committed:
style: исправлены замечания по ревью

EvgeniyaKuzmina committed:
style: исправлены замечания по ревью

Frilock reviewed (approved):
Принято
